config/resolver.go

```diff
@@ -138,45 +141,81 @@ func verifyRec(conf reflect.Value) error {
 	return nil
 }
 
-type Resolver interface {
-	SetEnvPrefix(prefix string) Resolver
-	SetConfigFile(filename string) Resolver
-	Resolve(config interface{}) Validator
+type Resolver[T any] interface {
+	SetEnvPrefix(prefix string) Resolver[T]
+	SetConfigFile(filename string) Resolver[T]
+	AddChangeCallback(func(*T)) Resolver[T]
+	Resolve(config *T) Validator
 }
 
-type configResolver struct {
-	Resolver
-	prefix     string
-	configFile string
+type configResolver[T any] struct {
+	Resolver[T]
+	prefix         string
+	configFile     string
+	watchCallbacks []func(*T)
 }
 
-func NewResolver() Resolver {
-	return &configResolver{}
+func NewResolver[T any]() Resolver[T] {
+	return &configResolver[T]{}
 }
 
-func (c *configResolver) SetEnvPrefix(prefix string) Resolver {
+func (c *configResolver[T]) SetEnvPrefix(prefix string) Resolver[T] {
 	c.prefix = prefix
 	return c
 }
 
 // SetConfigFile is the way to set the path to the configFile (including the name of the file)
-func (c *configResolver) SetConfigFile(filename string) Resolver {
+func (c *configResolver[T]) SetConfigFile(filename string) Resolver[T] {
 	c.configFile = filename
 	return c
 }
 
-func (c *configResolver) Resolve(config interface{}) Validator {
+func (c *configResolver[T]) AddChangeCallback(callback func(*T)) Resolver[T] {
+	c.watchCallbacks = append(c.watchCallbacks, callback)
+	return c
+}
+
+func (c *configResolver[T]) Resolve(config *T) Validator {
 	err := c.readFromFile(config)
 	if err == nil {
 		err = lamenv.Unmarshal(config, []string{c.prefix})
+		if len(c.watchCallbacks) != 0 {
+			c.watchFile(config)
+		}
 	}
 	return &validatorImpl{
 		err:    err,
 		config: config,
 	}
 }
 
-func (c *configResolver) readFromFile(config interface{}) error {
+func (c *configResolver[T]) watchFile(config *T) {
+	previousHash, _ := c.hashConfig(config)
+
+	err := file.Watch(c.configFile, func() {
+		err := c.readFromFile(config)
+		if err != nil {
+			logrus.WithError(err).Errorf("Cannot parse the watched config file %s", c.configFile)
```

Comment:
I find a bit opiniated to use logrus in a lib. @Nexucis Is it common in perses/common ?

Reply:
yes it is in the perses/common